File: frontend/src/lib/zod/magic-link.ts
/**
 * PR-CO-2-FE / PR-3E — Multi-action magic-link Zod schemas.
 *
 * Mirrors BE Pydantic models in `Backend_FastAPI/app/schemas/magic_link.py`:
 *   - MagicLinkAction enum (4 values, locked by DB CHECK)
 *   - MagicLinkConsumeRequest (cccd: 4 digits)
 *   - MagicLinkConsumeResponse (profile_id, action, status, consumed_at)
 *
 * Thin client per FRONTEND_ARCHITECTURE_V3.md: schemas are validation
 * only; no derivation, no business logic. Response shape trusted from
 * BE without re-deriving any state.
 */
import * as z from "zod"

/**
 * 4-action enum. Order matches BE for stable diffs; lowercase string
 * values match the DB CHECK constraint on `action_type`.
 */
export const magicLinkActionSchema = z.enum([
  "submit",
  "resubmit",
  "confirm",
  "withdraw",
])
export type MagicLinkAction = z.infer<typeof magicLinkActionSchema>

/**
 * Request body for POST /api/v2/admissions/magic-link/{action}/{token}.
 *
 * `cccd` is the last 4 digits of citizen ID (matches BE `^\d{4}$`
 * regex). NOT the full 12-digit CCCD — keeps UX consistent with the
 * existing /confirm/{token} 4-digit keypad.
 */
export const magicLinkConsumeRequestSchema = z.object({
  cccd: z
    .string()
    .regex(/^\d{4}$/u, "Vui lòng nhập đúng 4 chữ số"),
})
export type MagicLinkConsumeRequest = z.infer<
  typeof magicLinkConsumeRequestSchema
>

/**
 * Success response shape. `consumed_at` is ISO datetime string from BE
 * (Pydantic serialises `datetime` → ISO 8601).
 */
export const magicLinkConsumeResponseSchema = z.object({
  message: z.string(),
  profile_id: z.number().int().positive(),
  action: magicLinkActionSchema,
  status: z.string(),
  consumed_at: z.string(),
  next_url: z.string().nullable().optional(),
})
export type MagicLinkConsumeResponse = z.infer<
  typeof magicLinkConsumeResponseSchema
>

File: Backend_FastAPI/app/schemas/magic_link.py
"""PR-CO-2-BE / PR-3E: Pydantic schemas for the multi-action magic-link
consume endpoint.

Action enum locked to 4 values matching the DB CHECK constraint at
``AdmissionConfirmationToken.action_type`` (see
``alembic/versions/phase1_18_extend_confirmation_token_for_multi_action.py``).
"""
from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field


class MagicLinkAction(str, Enum):
    """4-action enum locked by DB CHECK at admission_confirmation_token."""
    SUBMIT = "submit"
    RESUBMIT = "resubmit"
    CONFIRM = "confirm"
    WITHDRAW = "withdraw"


class MagicLinkConsumeRequest(BaseModel):
    """POST body for /api/v2/admissions/magic-link/{action}/{token}.

    ``cccd`` is the last 4 digits of the candidate's CCCD (citizen ID).
    The service constant-time compares it against
    ``profile.citizen_id[-4:]`` via ``hmac.compare_digest``. Matches the
    existing single-action /confirm/{token} contract — keeps the FE
    keypad UX (4-digit) unchanged.
    """
    cccd: str = Field(
        ...,
        min_length=4,
        max_length=4,
        pattern=r"^\d{4}$",
        description="Last 4 digits of citizen ID",
    )


class MagicLinkConsumeResponse(BaseModel):
    """Response when a token is successfully consumed."""
    message: str
    profile_id: int
    action: MagicLinkAction
    status: str
    consumed_at: datetime
    next_url: Optional[str] = Field(
        default=None,
        description="Optional FE redirect target after success",
    )
